### PR TITLE
Input_embeds support

### PR DESCRIPTION
**Pull request by RinRin-32**
## Motivation

Adding input_embeds support for sglang, motivated and inspired from
https://github.com/sgl-project/sglang/issues/745

## Modifications

Changes in /manager to support new io_struct
Changes in /model_executor to support new io_struct
test cases under test/srt to demonstrate usage

In this implementation, curl request has to be made to use input_embeds.
This implementation DOES NOT have ARGS made to compliment input_embeds (I don't know where to make this edit),

When serving, to use input embeds please add 
`--disable-radix` tag
i.e.
`python3 -m sglang.launch_server --model-path Qwen/Qwen2.5-0.5B --port 30000 --disable-radix`

test_input_embeddings.py is built and tested on Qwen2.5-0.5B, please edit the model name according to the model that you'd like to test

input_embeddings.json is my test case embeddings for Qwen2.5-0.5B model and comparison_response.txt is a comparison between text based input and input_embeds input.

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

**Comment by merrymercy** [comment]
Thanks for the contribution. There is a related PR recently. Can you take a review on that? #2082

**Comment by RinRin-32** [comment]
> Thanks for the contribution. There is a related PR recently. Can you take a review on that? #2082

It seems that this PR update the documents and did more alteration with batch in mind. If their implementation doesn't require radix disabled then it'd be the perfect one among the 2 pr. My changes follow the guide in that open issue and is pretty minimal, with test to try out the feature.

Also can you please launch the checks again, I fixed some conditions since the last check you triggered, thank you.

File: python/sglang/srt/managers/schedule_batch.py [hunk 1]
```diff
@@ -177,6 +177,7 @@ def __init__(
         origin_input_ids: Tuple[int],
         sampling_params: SamplingParams,
         lora_path: Optional[str] = None,
+        input_embeds: Optional = None,
```
Comment (merrymercy) [inline python/sglang/srt/managers/schedule_batch.py]: Fill in the type annotation: Optional[what]?

File: python/sglang/srt/managers/schedule_batch.py [hunk 2]
```diff
@@ -442,6 +444,7 @@ class ScheduleBatch:
 
     # Batched arguments to model runner
     input_ids: torch.Tensor = None
+    input_embeds: Optional[Union[List[List[List[float]]], List[List[float]]]] = None
```
Comment (merrymercy) [inline python/sglang/srt/managers/schedule_batch.py]: can you make it a tensor at this stage?

File: python/sglang/srt/managers/schedule_batch.py [hunk 3]
```diff
@@ -1091,6 +1106,9 @@ class ModelWorkerBatch:
     # Sampling info
     sampling_info: SamplingBatchInfo
 
+    # The input Embeds
+    input_embeds: Optional[Union[List[List[List[float]]], List[List[float]]]] = None
```
Comment (merrymercy) [inline python/sglang/srt/managers/schedule_batch.py]: it is already a tensor here.

File: python/sglang/srt/managers/scheduler.py [hunk 4]
```diff
@@ -490,12 +490,26 @@ def handle_generate_request(
         self,
         recv_req: TokenizedGenerateReqInput,
     ):
+        # Check if input_embeds is present and create dummy input_ids
+        if recv_req.input_embeds is not None:
+            input_embeds_tensor = torch.tensor(recv_req.input_embeds)
```
Comment (merrymercy) [inline python/sglang/srt/managers/scheduler.py]: this conversion seems not to be used and it causes extra overhead.

File: python/sglang/srt/model_executor/forward_batch_info.py [hunk 5]
```diff
@@ -221,6 +224,11 @@ def init_new(
             global_num_tokens=batch.global_num_tokens,
             lora_paths=batch.lora_paths,
             sampling_info=batch.sampling_info,
+            input_embeds=(
+                batch.input_embeds.clone().detach().to(device)
```
Comment (merrymercy) [inline python/sglang/srt/model_executor/forward_batch_info.py]: remove this clone

File: test/srt/test_input_embeddings.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
+import json
+import os
+import subprocess
+import tempfile
+
+import torch
```
Comment (merrymercy) [inline test/srt/test_input_embeddings.py]: This looks good! Can you make it a unit test like the following examples
https://github.com/sgl-project/sglang/blob/main/test/srt/test_srt_endpoint.py
https://github.com/sgl-project/sglang/blob/main/test/srt/models/test_generation_models.py

File: python/sglang/srt/managers/tokenizer_manager.py [hunk 7]
```diff
@@ -195,9 +195,13 @@ async def _tokenize_one_request(
         obj: Union[GenerateReqInput, EmbeddingReqInput],
     ):
         """Tokenize one request."""
+        input_embeds = None
         # Tokenize
         input_text = obj.text
-        if obj.input_ids is None:
+        if obj.input_embeds is not None:
+            input_embeds = obj.input_embeds
+            input_ids = obj.input_ids
```
Comment (merrymercy) [inline python/sglang/srt/managers/tokenizer_manager.py:214]: we can assert `self.server_args.disable_radix_cache = False` here and return and raise a useful warning message.

**Review by merrymercy — CHANGES_REQUESTED** [review]
Thanks for the contribution. I feel your solution is better than the other one. We can try to merge yours! I left a few comments.

**Comment by merrymercy** [comment]
1. Can you fix the conflicts?
2. I added you as a co-author in this merged PR so you are not a first-time contributor anymore. The CI will be triggered automatically for any of your future commits. https://github.com/sgl-project/sglang/pull/2122

**Comment by RinRin-32** [comment]
I apologize for the slow reply.
Will get to it as soon as I can. Expect it done in a week!
